### DISCUSSION TITLE
Libre2 Direct normally OK but sometimes seems to be always disconnecting for hours

### DISCUSSION THREAD

**Discussion by nigelkeirross** [NightscoutFoundation/xDrip#3481] [Q&A]
In general, I have been successfully using Libre 2 direct (ie bridgeless) for ~ 18 months. However there appear to be times when it frequently disconnect & I miss readings.
When I am aware of this, I do an NFC scan, and often this will give an immediate NFC reading, and initiate the connection process. Often it will connect & then future readings will follow.
However, there are times when it seems to be disconnected all the time, or frequently disconnects, and most readings are missed. I then seem to need to do NFC scans every 5 minutes to keep the readings flowing.

Are there any settings that can help the connection persist, or be more stable?
Does anyone know what might put the system onto this "mood". In my experience it often happens when I am playing golf, or travelling on a (London underground) train - probably lots of BT activity close by. Almost the opposite of the golf situation. And sometimes while sitting at home it will be like this for an hour or two, and then go back to normal - where no missed readings for hours & hours.

I am using xDrip from 2023-05-27 on Google Pixel 5 phone on Android 14.
My settings are standards from the Libre2 Direct section of the read-the-docs that Patrick Sonnerat provided.

**Comment by PhilTU** [comment]
Check there are no power saving or battery management actions enabled on xDrip, Bluetooth or OOP2. Select Foreground mode on OOP2. Remove Librelink.

It does take a few minutes to reconnect, as a disconnected sensor only advertises every 2 minutes. Too many signals nearby can be an issue, if out in the open have the phone on the same side of ther sensor as there's nothing to reflect the signal back. Power output on L2 sensor is not great.

**Comment by nigelkeirross** [comment]
@PhilTU Thanks. xDrip & OOP2 were already exempt from Battery management/optimisation, but have now set OOP2 to "use foreground service" as well as "use service", which it was before. Should I untick "use service"? 
I cannot find how to explicitly ensure no power saving on Bluetooth on my Pixel 5 phone. Any instructions?

**Reply by PhilTU** [reply]
Use service turns it on, Use foreground service tells it to be in the foreground. I have both ticked but the first may be redundant if you use the second.

Can you see the Bluetooth app ? If you can select system apps or all apps you might see it in Settings / Apps, or search in apps for Bluetooth and scroll down to App info you can access like any other app, tap on Battery or whatever your flavour of Android offers to get to any power saving features.

Another way is to search Settings for Power savings and I have a list of exceptions I can tick the box there for Bluetooth.

**Comment by nigelkeirross** [comment]
@PhilTU It seems that changing OOP2 to foreground service has not made a difference. I finally worked out how to find the BT app. 
Settings / Apps then scroll to "App battery usage", then 3 dots, "show system". Can now scroll to Bluetooth. In that can change from "Optimised" to "Unrestricted". So will see if that help. I am playing golf this afternoon - so a strong test!
This is, however strange as I almost never have the problem overnight (my looping app alerts if it doesn't get a CGM withing 5 mins 10 secs of previous. And many nights I have extreme battery saver turned on (old phone battery), and yet still works OK. And yet yesterday the problem resurfaced when I was outside delivering leaflets for about 90 minutes. Almost always needing an NFC to reconnect. I always carry my phone in shirt top pocket, indoors & out.

**Reply by PhilTU** [reply]
hope it works. There isn't anything to go on when these things stop working so we have to clutch at straws !

**Reply by nigelkeirross** [reply]
Thanks. That is also my approach to golf! I have spent much time recently looking at the xDrip+ System Status/BT Device page, watching "BT Device" switch between "Disconnected", "Connecting" & "Connected". Often when disconnected, it says that the next connection attempt will be in > 2 minutes (and so miss next reading), or the time between each retry grows significantly. This is when I NFC. It would be nice to have this done automatically if in Libre2 direct mode.

**Reply by nigelkeirross** [reply]
@PhilTU Aargh, I have just had my first disconnect since changing the BT settings. The phone is on the desk next to me, being charged. So this suggests that there is something beyond BT power management on the phone going on.

**Comment by nigelkeirross** [comment]
@PhilTU Just very strange and disappointing yesterday. From the moment I started playing, and for the next 4.5 hours, it kept disconnecting, or refusing to connect. I was continually having to NFC after a missed reading, and it would seems to take forever to try & connect. In all that time I think I only got 2 or three "follow-on" readings - where I got a reading 5 mins after an NFC reading, meaning it had connected & stayed connected until the next reading. But never 2 in a row. Then, after the game, from when I went to the pub, 45 minutes in the car, then evening at home, it was back to normal - probably only 3 disconnections in that time. I wore the same shirt as before to keep the phone in the top pocket. Hard to understand the cause. Very frustrating, especially as it is important to have CGM readings while doing sustained exercise..

**Reply by PhilTU** [reply]
Yes, the OOP2 screen is the best indicator of data arriving. Once a sensor decides it lost connection it only advertises itself fleetingly every 2 minutes so can take 2,4,6 minutes to connect if an opportunity is missed.

**Comment by parapenT1sta** [comment]
What I can tell for mine experience, the Bluetooth settings avalible here for Libre 2 direct connection are not the best for my Samsung device: https://xdrip.readthedocs.io/en/latest/install/libre2/#manually-changing-settings

In fact I had enable some options and has been working flawless. Only once in a while I have to perfom "restart collector" in System Status but rarely.

**Reply by nigelkeirross** [reply]
@parapenT1sta Thanks. 
Do you know what settings you changed from the ones in the docs? Even though I have a Google Pixel5, I am desperate enough to try any settings that might prevent me having to go back to using a bridge ( Bubble )

BTW In the troubleshooting section of the docs 
https://xdrip.readthedocs.io/en/latest/troubleshoot/libre2direct/
This is the QR code for Samsung:
https://xdrip.readthedocs.io/en/latest/troubleshoot/libre2direct-qr-samsung/

this sets the following:
<boolean name="bluetooth_trust_autoconnect" value="false" />
  <boolean name="bluetooth_use_scan" value="false" />
    <boolean name="bluetooth_allow_background_scans" value="false" />
Did you try that?

**Reply by nigelkeirross** [reply]
@parapenT1sta For some reason that fragment did not display. 
"bluetooth_trust_autoconnect", "bluetooth_use_scan"  and "bluetooth_allow_background_scans" are all set to false.

**Reply by parapenT1sta** [reply]
After performing the changes, "Forget Device" and retry to scan Libre 2 again. Making sure that you have no battery optimization enabled.

Here I disabled: "Agressive Service Restart"
![Screenshot_20240529_153307_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/608650/1df9d86b-ec71-4860-937e-2db0ea3d1737)

Some options are enabled because before I was using Dexcom 5 and they aren't affecting Libre 2 Direct connection. Give them a try:
![Screenshot_20240529_153321_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/608650/70beb28a-c8d0-4ba4-aa2b-71fb61017e7f)

**Reply by nigelkeirross** [reply]
Thanks. So it looks from that that the BT changes you made from the docs were to tun on Dex BT Watchdog, and Companion BT. 
And you did not turn off "use scanning" and "Allow Background scans", as had been advised specifically for Samsung phones. 
Does that correspond to your memory? 
Also when you were having the frequent disconnection problems, was there any pattern when they occurred (eg for me it seems when I am outdoors)?

**Reply by parapenT1sta** [reply]
I'm using as it is... these are screenshots from my Samsung Z Fold 4.  As I said, it's working forme, I wasn't able to make it work with the settings shared online.

I don't have many disconnections even in public transportation or even in large events, like concerts. But I have issues, I forget device, restart the phone and then scan Libre again and starts working.

**Reply by nigelkeirross** [reply]
OK thanks. This is good timing as I have just changed sensors today & still have the old one on my right arm, so I have connected my backup phone with your changed settings to the old one, while my normal phone is now connected to the new sensor on my left arm. I will go outside and walk around and see if there are any differences... 
(I was just about to press send, and I see that the backup phone has just disconnected :-( )

**Reply by parapenT1sta** [reply]
What Android do you have on your backup phone? I'm running Android 14. And in your case you might have to make adjustments on the settings. 
Can you show how it's showing in the BT Device tab?

**Reply by nigelkeirross** [reply]
I am using Android 14 on both. (I always update the backup first to test xdrip & SightRemote both work on new versions). I will upload the BT screens for backup first.
![dback](https://github.com/NightscoutFoundation/xDrip/assets/21008320/f51b121e-d723-480b-96d7-0cf46397bca5)

**Reply by nigelkeirross** [reply]
This is from the main phone. These are the original settings without the 2 changes you suggested (which I made to the backup)
![dmain](https://github.com/NightscoutFoundation/xDrip/assets/21008320/56b89ccf-9941-448b-8ed4-1c8bd51c8af8)

**Reply by nigelkeirross** [reply]
But my walking experiment (trying to trigger disconnects on my main phone) was not successful and not fully conclusive, but I walked towards some local woods, for about 50 mins, with both phones, watching the BT device status pages. The main device remained connected all the time, while the backup phone disconnected 5 times. In all cases it reconnected fairly quickly (no NFC scans were done), and only one reading was missed (actually delayed by ~ 2 mins). This is the opposite of what I was hoping for.

**Reply by parapenT1sta** [reply]
I just noticed you didn't enabled Bluetooth Companion. For some reason?

**Reply by nigelkeirross** [reply]
have in the first image which was the backup, that was looking to be the same as yours. (the other change was the Dex BT Watchdog). Neither of these are turned on in my main phone, which should correspond to the read-the-docs settings.

**Comment by nigelkeirross** [comment]
@PhilTU Do you know if there is a way to have the disconnects & reconnects logged to the error log? This might help me see if infact when it appears to be working fine (no missed readings, say overnight), it is actually disconnecting, but quickly reconnecting each time. It might help me track down what is behind this strange behaviour.
